rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
+
+    /// the default time-to-live (TTL) for insertions (note that re-inserting a key
+    /// with any value will update the TTL to use the default_ttl)
+    ///
+    /// default: no TTL (insertions will remain until deleted)
+    default_ttl: Option<Duration>
+}
+```
+
+`Db#put_put_with_options` can optionally specify a timestamp and a row-level TTL at insertion time 
+by leveraging newly added fields in `WriteOptions`:
+
+```rust
+pub struct WriteOptions {
+    // ...
+   
+    /// the timestamp (in millis) at which this operation is considered to be executed - 
+    /// note that SlateDB does not automatically resolve out-of-order insertions
+    /// (if you need to retrieve the most recent semantic insertion, specify the
+    /// option on ReadOptions)
+    ///
+    /// default: uses the clock time configured in the DbOptions
+    pub timestamp_ms: Option<i64>,
+   
+    /// the time-to-live (ttl) for this insertion. If this insert overwrites an existing
+    /// database entry, the TTL for the most recent entry will be canonical.
+    ///
+    /// default: the TTL configured in DbOptions when opening a SlateDB session
+    pub ttl: Option<Duration>,
+}
+```
+
+`Db#get_with_options` can optionally specify: a grace period with which to continue searching the
+LSM tree for a more recent insertion. This allows you to specify an upper bound for what you expect
+out-of-order insertions to be.
+
+```rust
+pub struct ReadOptions {
+    // ...
+
+    /// if set to a positive number, SlateDB will not return the first matching row for
+    /// each key lookup and instead will continue to traverse the LSM tree until the next
+    /// matching key is older than the found key's timestamp + retrieval_grace_ms. Logically,
+    /// this is an "upper bound" on the magnitude of out-of-order insertions.
+    ///
+    /// Example: if there are three insertions [KeyA@100, KeyA@50, KeyA@150] that are each
+    /// in separate layers of the LSM tree, you will see the following behavior:
+    /// - if retrieval_grace_ms is <=0, KeyA@100 will be returned, and only the first SST
+    ///   will be scanned
+    /// - if retrieval_grace_ms is <=50, KeyA@100 will be returned after having scanned
+    ///   the SST that contains KeyA@50
+    /// - if retrieval_grace_ms is > 50 and <= 100, KeyA@150 will be returned after having
```

Comment:
Not sure I follow the `<= 100`. Wouldn't this mean it takes anything between timestamp 100 and timestamp 200 since the first timestamp is 100 (KeyA@100)? Is it 100 because the KeyA@50 moves the watermark back?

I'm having a tough time reasoning about this. 😅

Reply:
> Is it 100 because the KeyA@50 moves the watermark back

Yup that was my reasoning, though we can make it always compared to the first key we find if we want to simplify things. The reason I chose it this way is that if you find any two keys that are `grace_ms` apart, you've effectively seen the "max out of order" that you can see. In other words, if I've seen keyA@50 I should expect to never see anything newer than keyA@150 in deeper levels if my limit on grace is 100.